docs/architecture/adr-004-split-denomination-keys.md

```diff
@@ -0,0 +1,69 @@
+# ADR 004: Split Denomination Keys
+
+## Changelog
+
+- 2020-01-08: Initial version
+- 2020-01-09: Alterations to handle vesting accounts
+- 2020-01-14: Updates from review feedback
+
+## Context
+
+With permissionless IBC, anyone will be able to send arbitrary denominations to any other account. Currently, all non-zero balances are stored along with the account in an `sdk.Coins` struct, which creates a potential denial-of-service concern, as too many denominations will become expensive to load & store each time the account is modified. See issues [5467](https://github.com/cosmos/cosmos-sdk/issues/5467) and [4982](https://github.com/cosmos/cosmos-sdk/issues/4982) for additional context.
+
+Simply rejecting incoming deposits after a denomination count limit doesn't work, since it opens up a griefing vector: someone could send a user lots of nonsensical coins over IBC, and then prevent the user from receiving real denominations (such as staking rewards).
+
+## Decision
```

Comment:
Missing reference to the changes in the `supply` module. The supply module currently runs a `TotalSupply` invariant that iterates over the accounts' coins and compares it to the passive tracker of coins on the supply `Keeper`.

Reply:
Added subsection on supply module changes.